File: 2-Stack/5-car_fleet.cpp
#include <iostream>
#include <stack>
#include <vector>
#include <algorithm>

using namespace std;

int carFleet(int target, vector<int>& position, vector<int>& speed) {
    int len = position.size();
    if (len == 0) {
        return 0;
    }

    vector<pair<int, int>> cars;
    for (int i = 0; i < len; i++) {
        cars.push_back({position[i], speed[i]});
    }
    sort(cars.rbegin(), cars.rend()); // ordena a partir do carro com a menor distância até o target (maior distância do início)

    stack<double> fleets;

    for (int i = 0; i < len; i++) {
        double time = (double)(target - cars[i].first) / cars[i].second; // tempo = (distancia até o target) / (velocidade)

        if (fleets.empty() || time > fleets.top()) {
            fleets.push(time);
        }
    }
    return fleets.size();
}

// sem stack
int carFleet(int target, vector<int>& position, vector<int>& speed) {
    int len = position.size();
    if (len == 0) {
        return 0;
    }

    vector<pair<int, int>> cars;
    for (int i = 0; i < len; i++) {
        cars.push_back({position[i], speed[i]});
    }
    sort(cars.rbegin(), cars.rend()); // ordena a partir do carro com a menor distância até o target (maior distância do início)

    int numFleets = 0;
    double curMaxTime = 0.0;

    for (int i = 0; i < len; i++) {
        double time = (double)(target - cars[i].first) / cars[i].second; // tempo = (distancia até o target) / (velocidade)

        if (time > curMaxTime) {
            curMaxTime = time;
            numFleets++;
        }
    }
    return numFleets;
}

int main() {

}

File: PYTHON/2-Stack/5-car_fleet.py
from typing import List

def carFleet(target: int, position: List[int], speed: List[int]) -> int:
    cars = list(zip(position, speed)) # para agrupar a posição e a velocidade
    cars.sort(reverse=True) # ordem decrescente

    fleets = []

    for p, s in cars:
        time = (target - p) / s
        if not fleets or time > fleets[-1]:
            fleets.append(time)

    return len(fleets) 

if __name__ == "__main__":
    position = [4, 1, 0, 7]
    speed = [2, 2, 1, 1]
    target = 10
    print(carFleet(target, position, speed))